frontend/src/js/util/auth/authUploadWithProgress.ts

```diff
@@ -1,57 +1,152 @@
-import store from './../store';
-import { handleResponseFromAuthRequest } from './handleResponseFromAuthRequest';
-import { WorkspaceUploadMetadata } from '../../components/Uploads/UploadFiles';
+import store from "./../store";
+import { handleResponseFromAuthRequest } from "./handleResponseFromAuthRequest";
+import { WorkspaceUploadMetadata } from "../../components/Uploads/UploadFiles";
 
-export type ProgressHandler = (loadedBytes: number, totalBytes: number) => void
+export type ProgressHandler = (loadedBytes: number, totalBytes: number) => void;
+const RETRY_MAX_LIMIT = 3;
 
 export default function authUploadWithProgress(
-    url: string,
-    uploadId: string,
-    file: File,
-    path: string,
-    workspace?: WorkspaceUploadMetadata,
-    onProgress?: ProgressHandler
+  url: string,
+  uploadId: string,
+  file: File,
+  path: string,
+  workspace?: WorkspaceUploadMetadata,
+  onProgress?: ProgressHandler
 ) {
-    return new Promise((resolve, reject) => {
-        const xhr = new XMLHttpRequest();
-        xhr.upload.onprogress = (e) => {
-            if (e.lengthComputable) {
-                if(onProgress) {
-                    onProgress(e.loaded, e.total);
-                }
-            }
-        };
-
-        xhr.onloadend = () => {
-            handleResponseFromAuthRequest(
-                xhr.status,
-                xhr.getResponseHeader('X-Offer-Authorization')
-            );
-            if (xhr.status >= 200 && xhr.status < 300) {
-                resolve(xhr.response);
-            } else {
-                reject(`${xhr.status} - ${xhr.responseText}`);
-            }
-        };
-
-        xhr.open('POST', url);
-
-        // this isn't nice, but I don't want to get bogged down typing all the redux stuff
-        const state = store.getState() as any;
-        xhr.setRequestHeader('Authorization', 'Bearer ' + state.auth.jwtToken);
-
-        xhr.setRequestHeader('Content-Type', file.type);
-        xhr.setRequestHeader('Content-Location', encodeURIComponent(path));
-        // TODO MRB: this is a lot of headers which is a bit silly and also potentially too 'custom'
-        // for the more intrusive firewalls of the world. Uploading as FormData is probably cleaner
-        // but we'd need to check it's still fine for big'ish files
-        xhr.setRequestHeader('X-PFI-Upload-Id', uploadId);
-        if(workspace) {
-            xhr.setRequestHeader('X-PFI-Workspace-Id', workspace.workspaceId);
-            xhr.setRequestHeader('X-PFI-Workspace-Parent-Node-Id', workspace.parentNodeId);
-            xhr.setRequestHeader('X-PFI-Workspace-Name', workspace.workspaceName);
+  return new Promise((resolve, reject) => {
+    const xhr = new XMLHttpRequest();
+    const retryInitialCount = 0;
+
+    xhr.upload.onprogress = (e) => {
+      console.log(
+        `progress: loaded: ${e.loaded} from total: ${e.total} for url: ${url}`
+      );
```

Comment:
in my sample size of 1, there was some progress on the file (99%) before it appeared to stop, then there was a long pause until it hit the NS_ERROR_NET_INTERRUPT and failed before triggering the retry. If we wanted to trigger the retry sooner we could probably detect the stopped progress, though as this is a fairly intermittent bug I'm not sure it's worth the extra complexity - will just mean users have to wait the full minute before the retry kicks in

Reply:
Interesting point. During the long pause, were you still getting progress logs? I'm just wondering if the progress handler periodically get triggered no matter if there is any progress or not OR if the handler only gets triggered while there is progress on the upload

Reply:
No I didn't get any progress logs during the pause - it just appeared to stop, I think it only gets triggered for actual progress. So if we wanted to detect 'no progress' I think we'd have to use some kind of setTimeout which checks what the progress is and alarms if it hasn't changed. I think too much complexity to add at this point though! It would speed up the retry but hopefully there won't be that many errors